File: api/internal/store/resource_counts_test.go
package store

import (
	"testing"
	"time"

	"flow/api/internal/domain"
)

func TestRefreshResourceCountsUsesActiveRelationships(t *testing.T) {
	archivedAt := time.Now().UTC()
	issueLabel := domain.IssueLabel{ID: "issue-label", ResourceType: "issue", IssueCount: 99}
	projectLabel := domain.IssueLabel{ID: "project-label", ResourceType: "project", IssueCount: 99}
	initiativeLabel := domain.IssueLabel{ID: "initiative-label", ResourceType: "initiative", IssueCount: 99}
	projectSummary := &domain.ProjectSummary{ID: "project-active"}
	data := domain.Bootstrap{
		Labels: []domain.IssueLabel{issueLabel, projectLabel, initiativeLabel},
		Issues: []domain.Issue{
			{ID: "issue-active", Labels: []domain.IssueLabel{issueLabel, issueLabel}, Project: projectSummary},
			{ID: "issue-archived", Labels: []domain.IssueLabel{issueLabel}, Project: projectSummary, ArchivedAt: &archivedAt},
		},
		Projects: []domain.Project{
			{ID: "project-active", LabelIDs: []string{projectLabel.ID, projectLabel.ID}},
			{ID: "project-archived", LabelIDs: []string{projectLabel.ID}, ArchivedAt: &archivedAt},
		},
		Initiatives: []domain.Initiative{{ID: "initiative-active", LabelIDs: []string{initiativeLabel.ID}}},
	}

	refreshResourceCounts(&data)

	for _, label := range data.Labels {
		if label.IssueCount != 1 {
			t.Fatalf("label %s count = %d, want 1", label.ID, label.IssueCount)
		}
	}
	if data.Issues[0].Labels[0].IssueCount != 1 {
		t.Fatalf("embedded issue label count = %d, want 1", data.Issues[0].Labels[0].IssueCount)
	}
	if data.Projects[0].IssueCount != 1 || data.Projects[1].IssueCount != 0 {
		t.Fatalf("project issue counts = %d, %d, want 1, 0", data.Projects[0].IssueCount, data.Projects[1].IssueCount)
	}
}

File: web/src/lib/resource-counts.test.ts
import { describe, expect, it } from 'vitest'
import type { BootstrapData } from '@/types/flow'
import { deriveResourceCounts } from './resource-counts'

describe('resource counts', () => {
  it('counts active issue, project, and initiative relationships without duplicates', () => {
    const label = { id: 'label-1', name: 'Product', color: '#123456', resourceType: 'issue', issueCount: 99 }
    const projectLabel = { id: 'label-2', name: 'Portfolio', color: '#654321', resourceType: 'project', issueCount: 99 }
    const data = {
      labels: [label, projectLabel],
      issues: [
        { id: 'issue-1', labels: [label, label], project: { id: 'project-1' } },
        { id: 'issue-2', labels: [label], project: { id: 'project-1' }, archivedAt: '2026-08-01T00:00:00.000Z' },
      ],
      projects: [
        { id: 'project-1', labelIds: ['label-2', 'label-2'] },
        { id: 'project-2', labelIds: ['label-2'], archivedAt: '2026-08-01T00:00:00.000Z' },
      ],
      initiatives: [{ id: 'initiative-1', labelIds: ['label-2'] }],
    } as unknown as BootstrapData

    const result = deriveResourceCounts(data)
    expect(result.labels.find(item => item.id === 'label-1')?.issueCount).toBe(1)
    expect(result.labels.find(item => item.id === 'label-2')?.issueCount).toBe(2)
    expect(result.projects.find(item => item.id === 'project-1')?.issueCount).toBe(1)
    expect(result.projects.find(item => item.id === 'project-2')?.issueCount).toBe(0)
    expect(result.issues[0].labels[0].issueCount).toBe(1)
  })
})
